### PR TITLE
HTTP Metrics Path Matching

### PR DESCRIPTION
**Pull request by nelson-parente**
# Description

This PR implements the proposal https://github.com/dapr/proposals/pull/58. 

```yaml
spec:
  metric:
    enabled: true
    http:
      increasedCardinality: true
      pathMatching:
        ingress:
        - /orders/{orderID}/items/{itemID}
        - /users/{userID}
        egress:
        - /orders
        - /orders/{orderID}/items/{itemID}
```

PathMatching configuration is opt-in and ensures consistent representation of HTTP request paths in server metrics. 

## Issue reference

HTTP Metrics Path Matching Proposal: https://github.com/dapr/proposals/pull/58
Dapr Low Cardinality Issue with initial motivation for the proposal: https://github.com/dapr/dapr/issues/7719

Please reference the issue this PR will close: #_[issue number]_

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/diagnostics/http_monitoring.go
Comment (filintod) [inline pkg/diagnostics/http_monitoring.go:370]: should we also check `config.enabled == false` to return here as well?

Comment (filintod) [reply pkg/diagnostics/http_monitoring.go:370]: ok. I see you check later.

File: pkg/diagnostics/utils/path_normalization_utils.go
Comment (filintod) [inline pkg/diagnostics/utils/path_normalization_utils.go:20]: why we need this to be a pointer? I did not see a check against nil or something like it that would warrant it, but maybe I missed it. Asking as you already use pointers against the struct.

Comment (nelson-parente) [reply pkg/diagnostics/utils/path_normalization_utils.go:22]: I used the pointer to pass by reference instead of passing by value and with that, we can perform the path transformation by modifying the var.

Comment (nelson-parente) [reply pkg/diagnostics/utils/path_normalization_utils.go]: thanks for the tip! as you pointed out the issue was on the return value. Fixed in this commit: https://github.com/dapr/dapr/pull/7746/commits/8dc24c680ddb1b98d850a101cd258802a06101b4

File: pkg/apis/configuration/v1alpha1/types.go
Comment (ItalyPaleAle) [inline pkg/apis/configuration/v1alpha1/types.go:231]: No need for `*bool` since the default is false, just `bool`

Please don't forget `,omitempty` here and below

File: pkg/config/configuration.go
Comment (ItalyPaleAle) [inline pkg/config/configuration.go:280]: Couldn't this be just `return m.HTTP.PathNormalization`? Why creating a new object?

```suggestion
	if m.HTTP == nil {
		return nil
	}
	return m.HTTP.PathNormalization
```

File: pkg/diagnostics/http_monitoring.go
Comment (ItalyPaleAle) [inline pkg/diagnostics/http_monitoring.go:244]: Using `append` causes re-allocations in a hoth path. Best to create 2 separate slices for the enabled and disabled cases

File: pkg/diagnostics/http_monitoring.go
Comment (ItalyPaleAle) [inline pkg/diagnostics/http_monitoring.go:208]: Here too, avoid using `append` and just have 2 else's:

- `else if h.pathNormalization.enabled` and `else`

File: pkg/diagnostics/http_monitoring.go
Comment (ItalyPaleAle) [inline pkg/diagnostics/http_monitoring.go:443]: The proposal said

> [...] normalizing metrics without relying on regex's, which are known to be CPU-intensive.

However here you are:

1. Creating `http.Request` objects, which allocates a lot of resources
2. Invoking a HTTP server
3. Have the HTTP server do routing, which internally most likely relies on some regular expressions

If you do want to do this in a lightweight fashion, you should look into other options that don't involve all this work. This is a very hot path and it should be optimized.

Some ideas could include using a router only, possibly finding one that uses optimized trees and doesn't involve regexs.

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]
Normally the mainainters would tell you it's a bit premature to do the work before the proposal is even reviewed...!

**Comment by artursouza** [comment]
This proposal was reviewed among maintainers and people got invited to participate in the discussion too. While there is space for discussion, multiple maintainers thought this proposal to be reasonable and low risk to start implementation.

**Comment by artursouza** [comment]
@nelson-parente Please, address the optimization suggestions as this is a hot path for Dapr and can cause a performance regression.

Thanks, @ItalyPaleAle

**Comment by ItalyPaleAle** [comment]
> This proposal was reviewed among some runtime maintainers and people got invited to participate in the discussion too. While there is space for discussion, multiple maintainers thought this proposal to be reasonable and low risk to start implementation.

This is interesting as every time I was in @nelson-parente's shoes, the response I got from maintainers was very different. I have not seen voting happening (to my understanding, you will need the majority of 3 votes - 1 person for each company), and only 1 person commented.

**Comment by yaron2** [comment]
The PR is still pending proposal approval

File: pkg/diagnostics/http_monitoring.go
Comment (JoshVanL) [reply pkg/diagnostics/http_monitoring.go:443]: Yep- we should be using `http.Request` to instantiate the object :+1: 

We are not invoking any HTTP server here- we are invoking the mux router to hit a resulting handler. This mux doesn't use `regexp` which the reason why it was chosen.

File: pkg/diagnostics/http_monitoring.go
Comment (nelson-parente) [reply pkg/diagnostics/http_monitoring.go]: Since the tags are the same as in legacy mode instead of having 2 slices I added `|| h.pathnormalization.enabled` to the if.  Do you agree with this change?

File: pkg/apis/configuration/v1alpha1/types.go
Comment (nelson-parente) [reply pkg/apis/configuration/v1alpha1/types.go]: I think `*bool` would help us in case we want to change the default value, this way we always fallback to false.

Anyway, I removed it and added a test for the default case when pathNormalization is not in the spec.

File: pkg/diagnostics/http_monitoring.go
Comment (nelson-parente) [reply pkg/diagnostics/http_monitoring.go]: Changed the instantiation to use http.Request{...}.

As @filintod mentions we are not relying on regex evaluation here.

Given that we are registering the handlers once at startup and not invoking the HTTP server I think the solution using the mux routing primitives does not add much overhead other than the allocation of the request object.

File: pkg/config/configuration.go
Comment (ItalyPaleAle) [inline pkg/config/configuration.go:273]: This last nil-check for `m.HTTP.PathNormalization` is redundant

File: pkg/diagnostics/http_monitoring.go
Comment (ItalyPaleAle) [reply pkg/diagnostics/http_monitoring.go]: As long as it makes sense works for me!

File: pkg/diagnostics/http_monitoring.go
Comment (ItalyPaleAle) [reply pkg/diagnostics/http_monitoring.go]: Could you run some benchmarks please?

Comment (nelson-parente) [reply pkg/diagnostics/http_monitoring.go]: ofc! I will work on that!

File: pkg/diagnostics/http_monitoring.go
Comment (filintod) [inline pkg/diagnostics/http_monitoring.go:424]: you might want to benchmark with `sync.pool` to create this as well, to compare with/without it, as sync.pool might make sense in this use case

File: pkg/diagnostics/http_monitoring.go
Comment (nelson-parente) [reply pkg/diagnostics/http_monitoring.go]: **Benchmark Results**

All runs were done over this commit https://github.com/dapr/dapr/pull/7746/commits/1fccd567e0cd39e68b64267c14f7f7e49600cd52 (this will be reverted before merge)

_BenchmarkHTTPMiddlewareWithoutPathNormalization_

```
$  go test -benchtime=100x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithoutPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithoutPathNormalization-14    	     100	   5669265 ns/op	    4865 B/op	      85 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	0.951s

$ go test -benchtime=500x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithoutPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithoutPathNormalization-14    	     500	   5676510 ns/op	    4811 B/op	      84 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	3.162s

$ go test -benchtime=1000x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithoutPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithoutPathNormalization-14    	    1000	   5653330 ns/op	    4806 B/op	      84 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	5.963s
```

_BenchmarkHTTPMiddlewareWithPathNormalization_

```
$ go test -benchtime=100x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalization-14    	     100	   5673615 ns/op	    6440 B/op	     114 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	0.896s

$ go test -benchtime=500x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalization-14    	     500	   5717264 ns/op	    6342 B/op	     112 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	3.180s

$ go test -benchtime=1000x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalization-14    	    1000	   5652908 ns/op	    6331 B/op	     112 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	5.976s
```

_BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool_

```
$ go test -benchtime=100x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool-14    	     100	   5755573 ns/op	    6179 B/op	     112 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	0.902s

$ go test -benchtime=500x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool-14    	     500	   5721990 ns/op	    6049 B/op	     111 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	3.171s

$ go test -benchtime=1000x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool-14    	    1000	   5691965 ns/op	    6024 B/op	     111 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	6.022s
```

Comment (ItalyPaleAle) [reply pkg/diagnostics/http_monitoring.go]: The number of allocations went up by 1/3, which is quite a bit of added pressure to the GC.
Looks like the sync.Pool didn't help much as there was just 1 allocation saved.

Comment (ItalyPaleAle) [reply pkg/diagnostics/http_monitoring.go]: I am un-convinced that using the full HTTP request stack is particularly efficient. I don't have an exact number to give you as to what's an acceptable number of added allocations, but it'd be great to limit it.

Maybe consider using a third-party router to integrate? And just the router? https://github.com/julienschmidt/httprouter is supposed to be very fast with very low allocations, if the caveats aren't a concern.

Comment (nelson-parente) [reply pkg/diagnostics/http_monitoring.go]: @filintod noticed something in the test setup that is interesting which is the fact that since all tests were running in low cardinality the allocation increase is not related to the HTTP routing but to the fact that more tags are added in the requests with path normalization since these include path and method keys/values.

An issue was already opened about this: https://github.com/dapr/dapr/issues/6681

Given this, I ran the benchmarks again with high cardinality and the results with and without path normalization are very similar. This validates that the memory allocation increase observed in the first benchmark run was related to the fact that more tags are added and not to the path-matching implementation. 

_BenchmarkHTTPMiddlewareWithoutPathNormalization_

```
$ go test -benchtime=100x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithoutPathNormalization$ github.com/dapr/dapr/pkg/diagnostics

goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithoutPathNormalization-14    	     100	   5620625 ns/op	    7681 B/op	     126 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	0.938s

$ go test -benchtime=500x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithoutPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithoutPathNormalization-14    	     500	   5627427 ns/op	    7802 B/op	     125 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	3.136s

$ go test -benchtime=1000x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithoutPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithoutPathNormalization-14    	    1000	   5639435 ns/op	    7804 B/op	     126 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	5.949s
```

_BenchmarkHTTPMiddlewareWithPathNormalization_

```
$ go test -benchtime=100x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalization-14    	     100	   5641299 ns/op	    7121 B/op	     126 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	0.876s

$ go test -benchtime=500x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalization-14    	     500	   5667083 ns/op	    7018 B/op	     124 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	3.156s

$ go test -benchtime=1000x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalization-14    	    1000	   5670418 ns/op	    7007 B/op	     125 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	5.980s
```

_BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool_

```
$ go test -benchtime=100x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool$ github.com/dapr/dapr/pkg/diagnostics

goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool-14    	     100	   5673832 ns/op	    6857 B/op	     124 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	0.878s

$ go test -benchtime=500x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool$ github.com/dapr/dapr/pkg/diagnostics

goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool-14    	     500           5687737 ns/op	    6714 B/op	     122 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	3.160s

$ go test -benchtime=1000x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool$ github.com/dapr/dapr/pkg/diagnostics

goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool-14    	    1000	   5663766 ns/op	    6712 B/op	     123 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	5.983s
```

File: pkg/apis/configuration/v1alpha1/types.go
Comment (JoshVanL) [inline pkg/apis/configuration/v1alpha1/types.go:229]: Add API comments.

File: pkg/config/configuration_test.go
Comment (JoshVanL) [inline pkg/config/configuration_test.go:636]:
```suggestion
```

File: pkg/diagnostics/http_monitoring_benchmark_test.go
Comment (JoshVanL) [inline pkg/diagnostics/http_monitoring_benchmark_test.go:1]: Add copyright headers.

File: pkg/diagnostics/http_monitoring.go
Comment (JoshVanL) [inline pkg/diagnostics/http_monitoring.go:49]:
```suggestion
```

File: pkg/diagnostics/http_monitoring.go
Comment (JoshVanL) [inline pkg/diagnostics/http_monitoring.go:85]: Create a `patchMaching` struct containing only a `mux *http.ServeMux`. `httpMetrics` can then have 2 `patchMatching`s, ingress and egress.

File: pkg/diagnostics/http_monitoring.go
Comment (JoshVanL) [inline pkg/diagnostics/http_monitoring.go:194]: Why does having path matching rules enabled put us in the legacy code path?

File: pkg/diagnostics/http_monitoring.go
Comment (JoshVanL) [inline pkg/diagnostics/http_monitoring.go:350]:
```suggestion
func  newPathMatching(paths []string) *pathMatching {
```

File: pkg/diagnostics/http_monitoring.go
Comment (JoshVanL) [inline pkg/diagnostics/http_monitoring.go:385]: Rather than passing around strings and doing any/interface conversions- can we put the `stats.RecordWithTags` inside these request handlers?

**Review by JoshVanL — CHANGES_REQUESTED** [review]

File: pkg/diagnostics/http_monitoring.go
Comment (nelson-parente) [reply pkg/diagnostics/http_monitoring.go:194]: I changed that after this comment https://github.com/dapr/dapr/pull/7746#discussion_r1608424871

With path matching the keys used are the same as legacy code path the goal here was to simplify the code since these keys are not used in low cardinality.

Comment (nelson-parente) [reply pkg/diagnostics/http_monitoring.go]: path matching is orthogonal to low/high cardinality so we need to have this in the legacy code path to allow path matching in legacy (high cardinality)

File: pkg/diagnostics/http_monitoring.go
Comment (nelson-parente) [reply pkg/diagnostics/http_monitoring.go]: I don't think we can avoid the interface conversions here since we always need to cast the appropriate response writer 🙁 

Moving the metric producers inside the handlers would require us to have one rw per metric type. Personally, I'd prefer to keep things as they are now and not introduce that change in this PR if you are OK.

File: pkg/diagnostics/http_monitoring_path_matching.go
Comment (JoshVanL) [inline pkg/diagnostics/http_monitoring_path_matching.go:28]: Should we return `nil` here?

File: pkg/diagnostics/http_monitoring_path_matching.go
Comment (JoshVanL) [inline pkg/diagnostics/http_monitoring_path_matching.go:101]: Please put `const`s at the top of the file.

File: pkg/diagnostics/http_monitoring_path_matching.go
Comment (JoshVanL) [inline pkg/diagnostics/http_monitoring_path_matching.go:64]: Rather than have 2 `HandlerFunc` registrations, we can optionally append `"/"` if in non-legacy mode and remove dupes in the paths slice.


```go
if !legacy {
  paths = append(paths, "/")
  slices.Sort(paths)
  paths = slices.Compact(paths)
}
```


```suggestion
```

File: pkg/diagnostics/http_monitoring_path_matching.go
Comment (JoshVanL) [inline pkg/diagnostics/http_monitoring_path_matching.go:35]:
```suggestion
```

File: pkg/diagnostics/http_monitoring_path_matching.go
Comment (JoshVanL) [inline pkg/diagnostics/http_monitoring_path_matching.go:85]: Why is this required?

File: pkg/diagnostics/http_monitoring_path_matching.go
Comment (JoshVanL) [inline pkg/diagnostics/http_monitoring_path_matching.go:105]:
```suggestion
	matchedPath string
```

File: pkg/diagnostics/http_monitoring_path_matching.go
Comment (JoshVanL) [inline pkg/diagnostics/http_monitoring_path_matching.go:113]: `http.ResponseWriter` is a member of the struct so these should not be necessary. 

```suggestion
```

File: pkg/diagnostics/http_monitoring.go
Comment (JoshVanL) [inline pkg/diagnostics/http_monitoring.go:278]: Should a non-empty slice in one direction also enable the other? Is that what is happening?

**Review by JoshVanL — CHANGES_REQUESTED** [review]
Thanks @nelson-parente, couple more from me :slightly_smiling_face:

File: pkg/diagnostics/http_monitoring_path_matching.go
Comment (nelson-parente) [reply pkg/diagnostics/http_monitoring_path_matching.go:113]: The goal for these is that they serve as no-ops and don't do any actual work since we just want to have use routing and not the default rw mechanism. wdyt?

File: pkg/diagnostics/http_monitoring_path_matching.go
Comment (nelson-parente) [reply pkg/diagnostics/http_monitoring_path_matching.go:85]: Paths not starting `/` are invalid and the request will return an error on the handle function. 

I noticed this when testing and we had errors with `dapr/config`path.

File: pkg/diagnostics/http_monitoring.go
Comment (nelson-parente) [reply pkg/diagnostics/http_monitoring.go]: yeah, we can split each case 👍

File: pkg/diagnostics/http_monitoring_path_matching.go
Comment (nelson-parente) [reply pkg/diagnostics/http_monitoring_path_matching.go]: I think we need the 2 handlers registration because we want to distinguish when the user explicitly sets the "/" path or not.

When the user sets the "/" path we match it and use it otherwise if the user does not set a "/" we fallback to the unmatched placeholder "_" so we need to have different handler funcs for those cases.

File: pkg/diagnostics/http_monitoring_path_matching.go
Comment (nelson-parente) [reply pkg/diagnostics/http_monitoring_path_matching.go:75]: Update: all invoke method paths are coming without the initial slash. I fixed the dapr/config but all these others are using the `*invokev1.InvokeMethodRequest` and I don't really know the why these don't have the initial slash.

File: pkg/diagnostics/http_monitoring_path_matching.go
Comment (nelson-parente) [reply pkg/diagnostics/http_monitoring_path_matching.go]: done! we always need to register the catch all otherwise serveHTTP will throw an error when it doesn't match the path.

Comment (nelson-parente) [reply pkg/diagnostics/http_monitoring_path_matching.go]: done!

**Review by JoshVanL — APPROVED** [review]

**Review by artursouza — APPROVED** [review]

**Comment by marcduiker** [comment]
@holopin-bot @nelson-parente Enjoy your Dapr badge!

**Pull request merged**
